File: snippets/design-patterns/creational/singleton.go
// --- Singleton Pattern ---
// Use Case: Ensure a single instance of a struct exists across your application.
// Example: Configuration manager, database connection, or logger.
// Description: The Singleton Pattern restricts a struct to having only one instance
// and provides a global point of access to that instance.

package main

import (
	"fmt"
	"sync"
)

type Singleton struct {
	value string
}

var instance *Singleton
var once sync.Once

func GetInstance() *Singleton {
	once.Do(func() {
		instance = &Singleton{value: "I am the singleton instance"}
	})
	return instance
}

func main() {
	s1 := GetInstance()
	s2 := GetInstance()
	fmt.Println(s1 == s2) // Output: true
}

File: snippets/design-patterns/creational/singleton.py
'''
--- Singleton Pattern ---
Use Case: When you need to ensure a class has only one instance.
Example: configuration manager, logger, or a shared resource.
'''

class Singleton:
    _instance = None
    def __new__(cls, *args, **kwargs):
        if not cls._instance:
            cls._instance = super().__new__(cls, *args, **kwargs)
        return cls._instance
    

# Example Usage
singleton1 = Singleton()
singleton2 = Singleton()
assert singleton1 is singleton2  # True
